CRM/Civicase/Hook/ValidateForm/SendBulkEmail.php

```diff
@@ -0,0 +1,221 @@
+<?php
+
+/**
+ * Send bulk email.
+ */
+class CRM_Civicase_Hook_ValidateForm_SendBulkEmail {
+
+  /**
+   * Instance of the form sent on the hook.
+   *
+   * @var CRM_Core_Form
+   */
+  private $form;
+
+  /**
+   * Array with case roles to be notified.
+   *
+   * @var array|string[]
+   */
+  private $caseRoles;
+
+  /**
+   * Array of the case ids involved on the notification.
+   *
+   * @var array|string[]
+   */
+  private $caseIds;
+
+  /**
+   * Indicates of the client role is present on the roles to be notified.
+   *
+   * @var bool
+   */
+  private $isClientRoleSelected;
+
+  /**
+   * CRM_Civicase_Hook_ValidateForm_SendBulkEmail constructor.
+   *
+   * @throws CRM_Core_Exception
+   */
+  public function __construct() {
+    $caseRoles = CRM_Utils_Request::retrieve('caseRoles', 'String');
+    $this->caseRoles = $caseRoles ? explode(',', $caseRoles) : [];
+    $caseIds = CRM_Utils_Request::retrieve('allCaseIds', 'String');
+    $this->caseIds = $caseIds ? explode(',', $caseIds) : [];
+
+    $clientRoleKey = array_search('client', $this->caseRoles, TRUE);
+    $this->isClientRoleSelected = $clientRoleKey !== FALSE;
+    if ($this->isClientRoleSelected) {
+      unset($this->caseRoles[$clientRoleKey]);
+    }
+  }
+
+  /**
+   * Implement the send bulk functionality.
+   *
+   * @param string $formName
+   *   Form Name.
+   * @param array $fields
+   *   Fields List.
+   * @param array $files
+   *   Files list.
+   * @param CRM_Core_Form $form
+   *   Form Class object.
+   * @param array $errors
+   *   Errors.
+   *
+   * @return bool
+   *   TRUE if the hook ran, false otherwise.
+   */
+  public function run($formName, array &$fields, array &$files, CRM_Core_Form &$form, array &$errors) {
+    $this->form = $form;
+    if (!$this->shouldRun()) {
+      return FALSE;
+    }
+
+    $uniqueContacts = count($this->form->_contactIds);
+
+    $casesContactInfo = $this->getCasesContactInfo();
+    $messageSentCount = 0;
+    foreach ($casesContactInfo as $caseId => $contactIds) {
+      $this->sendEmailForCase($caseId, $contactIds);
+
+      $messageSentCount += count($contactIds);
+    }
+
+    if ($messageSentCount > $uniqueContacts) {
```

Comment:
Before this changes, if there was three contacts selected, then the same number of messages would have been sent. Now, if there are two cases selected behind, the message will be sent 3 x 2 = 6 times.
The status message is already set to `3 messages were sent...`, and then we need to modify it. I check if the replacement is necessary or not (it could be the case in which the numbers are the same).
By the way, I did a fix that I discovered during this fixes: https://github.com/compucorp/uk.co.compucorp.civicase/pull/804/commits/6626a46c970db95b53a295d5fef0e22ea35b4c73 In one case it was not correctly replaced.

Reply:
Not sure about the new logic of `3 x 2 = 6 times`.
Say for example two cases were selected for emails to be sent to the case role `Case coordinator`. If there were two contacts qualified for this, one for Case A and the other for Case B. Then, for this case there will only be two emails sent one per contact for each case rather than sending two emails for each of the contacts to make 4. 

However there is another scenario whereby the two cases are selected for emails to be sent to the case role `Case coordinator` and it happens that a single contact meets the criteria for both case. In this case two emails will still be sent, one per contact for each of the case. 

So I don't really understand the logic behind sending emails for all the contacts for each case. We should only send emails for contacts based on their relationship with the case.

Reply:
@tunbola Yes, my explanation was incorrect, is not the product of the number of contacts and number of cases.
Anyway, the code keeps being correct. For your first example, let's say that contact A is case coordinator of case A and contact B the case coordinate for case B. Then, the array is:
[
  Case A => [Contact A],
  Case B => [Contact B],
]
It is a total of two, then is not necessary to change the status.
Nos let's suppose than another contact is Benefit Specialist of Case A and also of Case B. Then, the array is:
[
  Case A => [Contact A, Contact C],
  Case B => [Contact B, Contact C],
]
Then, the code gives 2+2 = 4 emails. Before this change, there were 3 emails sent instead:
Contact A received case A info
Contact B received case A info
Contact C received case A info
The status needs to be modified.
That has more sense now?